src/lazy_es.rs

```diff
@@ -0,0 +1,168 @@
+        match self {
+            Self::NeedEsQuery { header, query, .. } => Some((header, query)),
+            _ => None,
+        }
+    }
+
+    /// Chain some computation out of the value which will eventually be
+    /// computed.
+    pub fn map<U>(self, func: impl FnOnce(T) -> U + 'p) -> LazyEs<'p, U> {
+        self.then(move |x| LazyEs::Value(func(x)))
+    }
+
+    /// Chain some lazy computation out of the value which will eventually be
+    /// computed. This means that one more elasticsearch request may be
+    /// required to compute the final result.
+    pub fn then<U>(self, func: impl FnOnce(T) -> LazyEs<'p, U> + 'p) -> LazyEs<'p, U> {
+        match self {
+            Self::Value(x) => func(x),
+            Self::NeedEsQuery {
+                header,
+                query,
+                progress,
+            } => LazyEs::NeedEsQuery {
+                header,
+                query,
+                progress: Box::new(move |val| progress(val).then(func)),
+            },
+        }
+    }
+
+    /// Send a request to elasticsearch to make progress for all computations
+    /// in `partials` that are not done yet.
+    pub fn batch_make_progress(rubber: &mut Rubber, partials: &mut [Self], max_batch_size: usize) {
+        let need_progress: Vec<_> = partials
+            .iter_mut()
+            .filter(|partial| partial.value().is_none())
+            .take(max_batch_size)
+            .collect();
+
+        let body: String = {
+            need_progress
+                .iter()
+                .filter_map(|partial| {
+                    let (header, query) = partial.header_and_query()?;
+                    Some(format!("{}\n{}\n", header.to_string(), query.to_string()))
+                })
+                .collect()
+        };
+
+        let es_response = rubber
+            .post("_msearch", &body)
+            .expect("ES query failed")
+            .text()
+            .expect("failed to read ES response");
+
+        let responses =
+            parse_es_multi_response(&es_response).expect("failed to parse ES responses");
+        assert_eq!(responses.len(), need_progress.len());
+
+        for (partial, res) in need_progress.into_iter().zip(responses) {
+            match partial {
+                LazyEs::NeedEsQuery { progress, .. } => {
+                    let progress = std::mem::replace(progress, Box::new(|_| unreachable!()));
+                    *partial = progress(res);
+                }
+                LazyEs::Value(_) => unreachable!(),
+            }
+        }
+    }
+
+    /// Run all input computations until they are finished and finally output
+    /// the resulting values.
+    pub fn batch_make_progress_until_value(
+        rubber: &mut Rubber,
+        mut partials: Vec<Self>,
+        max_batch_size: usize,
+    ) -> Vec<T> {
+        while partials.iter().any(|x| x.value().is_none()) {
+            Self::batch_make_progress(rubber, partials.as_mut_slice(), max_batch_size);
+        }
```

Comment:
I actually about writing it this way, I'll make it this way, it has to be clearer this way.

Even after this change there might be a way to build a task that never ends (by indefinitely querying ES), but I think it would be quite hard to write unintentionally.